src/commands/cmd_zset.cc

```diff
@@ -220,64 +222,218 @@ class CommandZLexCount : public Commander {
   RangeLexSpec spec_;
 };
 
-class CommandZPop : public Commander {
+class CommandZPop : public Commander,
+                    private EvbufCallbackBase<CommandZPop, false>,
+                    private EventCallbackBase<CommandZPop> {
  public:
-  explicit CommandZPop(bool min) : min_(min) {}
+  explicit CommandZPop(bool min, bool block) : min_(min), block_(block) {}
 
   Status Parse(const std::vector<std::string> &args) override {
-    if (args.size() > 3) {
-      return {Status::RedisParseErr, errWrongNumOfArguments};
-    }
+    if (!block_) {
+      if (args.size() > 3) {
+        return {Status::RedisParseErr, errWrongNumOfArguments};
+      }
+
+      if (args.size() == 3) {
+        auto parse_result = ParseInt<int>(args[2], 10);
+        if (!parse_result) {
+          return {Status::RedisParseErr, errValueNotInteger};
+        }
 
-    if (args.size() == 3) {
-      auto parse_result = ParseInt<int>(args[2], 10);
-      if (!parse_result) {
-        return {Status::RedisParseErr, errValueNotInteger};
+        count_ = *parse_result;
       }
+      keys_.push_back(args[1]);
+      return Commander::Parse(args);
+    }
+    auto parse_result = ParseInt<int>(args[args.size() - 1], 10);
 
-      count_ = *parse_result;
+    if (!parse_result) {
+      return {Status::RedisParseErr, "timeout is not an integer or out of range"};
     }
+
+    if (*parse_result < 0) {
+      return {Status::RedisParseErr, errTimeoutIsNegative};
+    }
+
+    timeout_ = *parse_result;
+
+    keys_ = std::vector<std::string>(args.begin() + 1, args.end() - 1);
     return Commander::Parse(args);
   }
 
   Status Execute(Server *svr, Connection *conn, std::string *output) override {
-    redis::ZSet zset_db(svr->storage, conn->GetNamespace());
-    std::vector<MemberScore> member_scores;
-    auto s = zset_db.Pop(args_[1], count_, min_, &member_scores);
+    svr_ = svr;
+    conn_ = conn;
+
+    auto s = TryPopFromMultiZset();
```

Comment:
The function `TryPopFromMultiZset()` is also used in function `OnWrite` later, i think expanding the code here rather than using a function will cause duplication.

Reply:
There may be some complexity because i merged zpopmin/zpopmax/bzpopmin/bzpopmax commands together.

Reply:
@Yangsx-1 I appreciate your effort to remove duplication - it's the right way to write clean code. However, right now, as I mentioned before, the `TryPopFromMultiZset` pops and sends only an error reply and stores popped elements in a class member variable - a lot of actions just for a single function. Maybe, at least, it can return status and then you will check it and take appropriate actions?

Reply:
Ok, i understand your opinion. I'll revise it.

Reply:
@Yangsx-1 Personally, I don't like functions with the `Try` prefix. 
Does the function `ReplyForBZPop` have something specific for blocking pop from a sorted set that you named it in this way?

Reply:
I just refer to BLPOP/BRPOP in file cmd_list.cc, so i used the Try prefix. Is there a better way to name it?
The reply of BZPOPMIN/BZPOPMAX has an additional 'key' compared to ZPOPMIN/ZPOPMAX. I think i can merge them together so that they can use the same function.